ffi/tests/ffi_export.rs

```diff
@@ -308,28 +338,100 @@ fn return_result() {
     unsafe {
         assert_eq!(
             FfiReturn::ExecutionFail,
-            FfiStruct__fallible_int_output(From::from(false), output.as_mut_ptr())
+            OpaqueStruct__fallible_int_output(From::from(false), output.as_mut_ptr())
         );
         assert_eq!(0, output.assume_init());
         assert_eq!(
             FfiReturn::Ok,
-            FfiStruct__fallible_int_output(From::from(true), output.as_mut_ptr())
+            OpaqueStruct__fallible_int_output(From::from(true), output.as_mut_ptr())
         );
         assert_eq!(42, output.assume_init());
     }
 }
 
+#[test]
+fn primitive_conversion() {
+    let byte: u8 = 1;
+    let mut output = MaybeUninit::new(0);
+
+    unsafe {
+        assert_eq!(
+            FfiReturn::Ok,
+            __freestanding_with_primitive(byte.into_ffi(), output.as_mut_ptr())
+        );
+
+        let ret_val = output.assume_init();
+        #[cfg(feature = "wasm")]
+        let ret_val: u8 =
+            TryFromReprC::try_from_repr_c(ret_val, &mut ()).expect("Conversion failed");
+
+        assert_eq!(1, ret_val);
+    }
+}
+
+#[test]
+fn fieldless_enum_conversion() {
+    let fieldless_enum = FieldlessEnum::A;
+    let mut output = MaybeUninit::new(2);
+
+    unsafe {
+        assert_eq!(
+            FfiReturn::Ok,
+            __freestanding_with_fieldless_enum(fieldless_enum.into_ffi(), output.as_mut_ptr())
+        );
+
+        let ret_val = TryFromReprC::try_from_repr_c(output.assume_init(), &mut ());
+        assert_eq!(FieldlessEnum::A, ret_val.expect("Conversion failed"));
+    }
+}
+
+#[test]
+fn data_carrying_enum_conversion() {
+    let data_carrying_enum = DataCarryingEnum::A(get_new_struct());
+    let mut output = MaybeUninit::new(DataCarryingEnum::B(42));
+
+    unsafe {
+        assert_eq!(
+            FfiReturn::Ok,
+            __freestanding_with_data_carrying_enum(data_carrying_enum.into_ffi(), output.as_mut_ptr())
```

Comment:
this is a problem because `into_ffi` produces `__iroha_ffi__ReprCDataCarryingEnumIntoFfiTarget` but `__freestanding_with_data_carrying_enum` takes `__iroha_ffi__ReprCDataCarryingEnumTryFromReprCSource`

I believe this could be resolved if we made `IntoFfi::into_ffi` also take `Store` as does `TryFromReprC`

Reply:
Explicit ownership over memory, right?